src/tools/chip-cert/CertUtils.cpp

```diff
@@ -0,0 +1,826 @@
+                chipAttrLen = 8;
+            }
+
+            if (!X509_NAME_add_entry_by_NID(X509_get_subject_name(cert), attrNID, MBSTRING_UTF8, (unsigned char *) chipAttrStr,
+                                            chipAttrLen, -1, 0))
+            {
+                ReportOpenSSLErrorAndExit("X509_NAME_add_entry_by_NID", res = false);
+            }
+        }
+        else
+        {
+            if (!X509_NAME_add_entry_by_NID(X509_get_subject_name(cert), attrNID, MBSTRING_UTF8,
+                                            (unsigned char *) rdn[i].mAttrValue.mString.mValue,
+                                            (int) rdn[i].mAttrValue.mString.mLen, -1, 0))
+            {
+                ReportOpenSSLErrorAndExit("X509_NAME_add_entry_by_NID", res = false);
+            }
+        }
+    }
+
+exit:
+    return res;
+}
+
+bool ToolChipDN::HasAttr(chip::ASN1::OID oid) const
+{
+    uint8_t rdnCount = RDNCount();
+
+    for (uint8_t i = 0; i < rdnCount; i++)
+    {
+        if (oid == rdn[i].mAttrOID)
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void ToolChipDN::PrintDN(FILE * file, const char * name) const
+{
+    uint8_t rdnCount = RDNCount();
+    char valueStr[128];
+    const char * certDesc = nullptr;
+
+    fprintf(file, "%s: ", name);
+
+    for (uint8_t i = 0; i < rdnCount; i++)
+    {
+        if (IsChip64bitDNAttr(rdn[i].mAttrOID))
+        {
+            snprintf(valueStr, sizeof(valueStr), "%016" PRIX64, rdn[i].mAttrValue.mChipVal);
+        }
+        else if (IsChip32bitDNAttr(rdn[i].mAttrOID))
+        {
+            snprintf(valueStr, sizeof(valueStr), "%08" PRIX32, static_cast<uint32_t>(rdn[i].mAttrValue.mChipVal));
+        }
+        else
+        {
+            uint32_t len = rdn[i].mAttrValue.mString.mLen;
+            if (len > sizeof(valueStr) - 1)
+            {
+                len = sizeof(valueStr) - 1;
+            }
+            memcpy(valueStr, rdn[i].mAttrValue.mString.mValue, len);
+            valueStr[len] = 0;
+        }
+
+        fprintf(file, "%s=%s", chip::ASN1::GetOIDName(rdn[i].mAttrOID), valueStr);
+        if (certDesc != nullptr)
+        {
+            fprintf(file, " (%s)", certDesc);
+        }
+    }
+}
+
+static bool SetCertSerialNumber(X509 * cert)
+{
+    bool res = true;
+    uint64_t rnd;
+    ASN1_INTEGER * snInt = X509_get_serialNumber(cert);
+
+    // Generate a random value to be used as the serial number.
+    if (!RAND_bytes(reinterpret_cast<uint8_t *>(&rnd), sizeof(rnd)))
+    {
+        ReportOpenSSLErrorAndExit("RAND_bytes", res = false);
+    }
+
+    // Avoid negative numbers.
+    rnd &= 0x7FFFFFFFFFFFFFFF;
+
+    // Store the serial number as an ASN1 integer value within the certificate.
+    if (ASN1_INTEGER_set_uint64(snInt, rnd) == 0)
+    {
+        ReportOpenSSLErrorAndExit("ASN1_INTEGER_set_uint64", res = false);
+    }
+
+exit:
+    return res;
+}
+
+static bool SetCertTimeField(ASN1_TIME * s, const struct tm & value)
```

Comment:
Can we choose a more descriptive name than `s` here?

Reply:
done